### ISSUE TITLE
Remove old sdk can't remove all old runtime

### ISSUE THREAD

**Issue by livarcocc** [dotnet/core-setup#4951]
_From @Xarkam on January 9, 2019 14:46_

Hello,

Today, I removed old sdk from my machine.
But, I have two old runtime not removed:
![image](https://user-images.githubusercontent.com/331816/50906063-93a5b700-1424-11e9-88e6-5e40f2bf028e.png)
![image](https://user-images.githubusercontent.com/331816/50906249-fa2ad500-1424-11e9-976f-a190e4af12a4.png)
![image](https://user-images.githubusercontent.com/331816/50906311-1c245780-1425-11e9-964d-7b4c466a66e7.png)
Is there a reason for these two versions to still be present?
Why folders 2.1.0 and 2.1.2 in `C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App` are not empty and is not listed as installed ?

How to delete them?

My sdk :
![image](https://user-images.githubusercontent.com/331816/50906146-c485ec00-1424-11e9-9e06-87cf022b91cf.png)

I do not have a runtime installed outside of those installed by the sdk.

Thanks

_Copied from original issue: dotnet/sdk#2804_

**Comment by livarcocc** [comment]
Not sure why 2.1.0 and 2.1.2 were not removed. You can remove them, at least according to the version of the SDK that you have installed. Do you happen to have logs for the unsinstalls? Otherwise, I don't see much we can do to address this and I will just close the issue.

**Comment by livarcocc** [comment]
_From @Xarkam on January 9, 2019 18:26_

If I remove 2.1.0 and 2.1.1 folders, a dotnet --info give me is always installed.
Where I can find logs files ?

**Comment by livarcocc** [comment]
We would need logs for the MSI uninstall process. @johnbeisner would you know where to find those?

> If I remove 2.1.0 and 2.1.1 folders, a dotnet --info give me is always installed.

I don't know what you mean by those. Are you saying you removed the folders and dotnet --info is still reporting them?

**Comment by livarcocc** [comment]
_From @Xarkam on January 10, 2019 7:54_

If I delete `C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App\2.1.0` and `C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App\2.1.2 folders`, a `dotnet --info` displays in the list that they are still installed.

**Comment by livarcocc** [comment]
Can you please paste the contents of `dotnet --info` after removing the folders?

**Comment by livarcocc** [comment]
_From @Xarkam on January 10, 2019 18:19_

This the content after reset my powershell session **and** after deleted folder.
```
SDK .NET Core (reflétant tous les global.json) :
 Version:   2.2.102
 Commit:    96ff75a873

Environnement d'exécution :
 OS Name:     Windows
 OS Version:  10.0.17763
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.2.102\

Host (useful for support):
  Version: 2.2.1
  Commit:  878dd11e62

.NET Core SDKs installed:
  2.1.202 [C:\Program Files\dotnet\sdk]
  2.1.503 [C:\Program Files\dotnet\sdk]
  2.2.102 [C:\Program Files\dotnet\sdk]

.NET Core runtimes installed:
  Microsoft.AspNetCore.All 2.1.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]

To install additional .NET Core runtimes or SDKs:
  https://aka.ms/dotnet-download
```

So, if you remove folder and you have cmd/powershell opened, the deletion is is not taken into account.
You must restart cmd/powershell session.

I suppose, I can make same for 2.1.202 sdk ? (not present in previous uninstaller list screenshot)

**Comment by livarcocc** [comment]
I am going to move this issue to core-setup, where the code to list the runtimes exists. I am not sure why you would have to restart powershell for this to work. It should pretty much just list folder relative to dotnet.exe itself.

**Comment by johnbeisner** [comment]
@livarcocc 
MSI uninstall logs [about 4-5 logfiles] will be in the %temp% folder
With names like: 
"Microsoft_.NET_Core_SDK_2.1.500_(x64)_20180110190429.log"
"Microsoft_.NET_Core_SDK_2.1.500_(x64)_20180110190429_004_dotnet_runtime_2.1.0_win_x64.msi.log"

**Comment by dleeapho** [comment]
#4188 should address this going forward. We are also considering tooling for cleaning up old installs. /cc @leecow @KathleenDollard

**Comment by dagood** [comment]
I see two issues here:

---

1: Why did `dotnet --info` list the old ASP.NET Core shared frameworks even after their dirs were deleted? 

Unfortunately, this didn't repro for me with a `2.1.8` host.

---

2: Why did uninstalling old SDKs not uninstall the old ASP.NET Core shared frameworks?

I don't know if we can be sure about this specific case without the logs. I can repro the behavior, though, by opening one of the DLLs then running the uninstaller. It doesn't remove the dir, and `dotnet --info` still lists it. (The files *in* the directory are removed, however.)

The AspNetCoreSharedFramework uninstall log says `DEBUG: Error 2911:  Could not remove the folder C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App\3.0.0-preview3-19120-02\.`

This isn't something the current plan for https://github.com/dotnet/core-setup/issues/4188 will deal with--it's tracking in-place upgrade. We could consider adding "bad uninstall" cleanup logic to a CLI tool that does more generic cleanup, however.

**Comment by rdzidziguri** [comment]
![image](https://user-images.githubusercontent.com/6897869/56647405-7377cc00-6692-11e9-9c4b-69a9245b2831.png)

Same problem ehre old runtimes left after uninstalling SDK.s

**Comment by dagood** [comment]
@rdzidziguri, do you have any of the log files mentioned above?

> MSI uninstall logs [about 4-5 logfiles] will be in the %temp% folder
> With names like:
> 
> ```
> "Microsoft_.NET_Core_SDK_2.1.500_(x64)_20180110190429.log"
> "Microsoft_.NET_Core_SDK_2.1.500_(x64)_20180110190429_004_dotnet_runtime_2.1.0_win_x64.msi.log"
> ```

Also:

* Are you sure the extra ASP.NET Core runtimes weren't installed manually? (Do they show up in Programs and Features?)
* Can you check to see if the extra directories are empty, or actually have shared framework files in them?

The extras look to me like Microsoft.AspNetCore.App/All 2.1.0, 2.1.1, and 2.1.2. Just to confirm that we're talking about the same ones.

**Comment by rdzidziguri** [comment]
@dagood 
Let me go through your questions one by one.

> do you have any of the log files mentioned above?

providing everything I have for now as an attachment [install_logs.zip](https://github.com/dotnet/core-setup/files/3113553/install_logs.zip)

> Are you sure the extra ASP.NET Core runtimes weren't installed manually? (Do they show up in Programs and Features?)

No, I did not install any there and they do not show up in add remove features 

> Can you check to see if the extra directories are empty, or actually have shared framework files in them?

They are not empty and they have files in them.

**Comment by dagood** [comment]
Thanks for the info!

From the log files, you uninstalled these and they went fine:

* SDK `2.1.202`, with runtime `2.0.9`
* SDK `2.1.503`, with runtime `2.1.7`
* SDK `2.1.504`, with runtime `2.1.8`
* SDK `2.2.103`, with runtime `2.2.1`
* SDK `2.2.105`, with runtime `2.2.3`

Question now is why your machine has the `2.1.0`/`2.1.1`/`2.1.2` runtimes.... Maybe you (or something--Visual Studio?) tried to uninstall it a while back and that had a bug, so now we don't have a record of it.

I tried installing and uninstalling [SDK 2.1.302](https://dotnet.microsoft.com/download/dotnet-core/2.1) which carries runtime `2.1.2` to see if there's some simple repro, but the uninstall removed the files as expected.

I think livarcocc's reply from earlier applies here:

> Not sure why 2.1.0 and 2.1.2 were not removed. You can remove them, at least according to the version of the SDK that you have installed. Do you happen to have logs for the unsinstalls? Otherwise, I don't see much we can do to address this and I will just close the issue.

The uninstall issue itself seems to not be present in recent SDKs, so it *seems* fixed although we don't know for sure what the original problem is. I think any effort to help clean up old broken uninstalls would be handled by a separate tool, if we go after it:

> This isn't something the current plan for #4188 will deal with--it's tracking in-place upgrade. We could consider adding "bad uninstall" cleanup logic to a CLI tool that does more generic cleanup, however.

Closing--I think we'll just have to continue to suggest the workaround of deleting the directories until we have a tool that can detect/fix this automatically or get a report of this issue while the uninstall bug is still "alive" in the logs.
